letsencrypt/acme/jose.py

```diff
@@ -0,0 +1,100 @@
+"""JOSE."""
+import base64
+import binascii
+
+import Crypto.PublicKey.RSA
+
+from letsencrypt.acme import util
+
+
+def _leading_zeros(arg):
+    if len(arg) % 2:
+        return '0' + arg
+    return arg
+
+
+class JWK(util.JSONDeSerializable, util.ImmutableMap):
+    # pylint: disable=too-few-public-methods
+    """JSON Web Key.
+
+    .. todo:: Currently works for RSA public keys only.
+
+    """
+    __slots__ = ('key',)
+    schema = util.load_schema('jwk')
+
+    @classmethod
+    def _encode_param(cls, param):
+        """Encode numeric key parameter."""
+        return b64encode(binascii.unhexlify(
+            _leading_zeros(hex(param)[2:].rstrip('L'))))
+
+    @classmethod
+    def _decode_param(cls, param):
+        """Decode numeric key parameter."""
+        return long(binascii.hexlify(b64decode(param)), 16)
+
+    def to_json(self):
+        """Serialize to JSON."""
+        return {
+            'kty': 'RSA',  # TODO
+            'n': self._encode_param(self.key.n),
+            'e': self._encode_param(self.key.e),
+        }
+
+    @classmethod
+    def _from_valid_json(cls, jobj):
```

Comment:
Intention was to force API clients to use `from_json` instead (defined in `util.JSONDeSerializable`), which makes sure input validates against JSON schema before calling `_from_valid_json`.

However, I've changed it in my local development `acme.challenges` branch, where it's no longer the case that subobjects of ACME message have `schema` attribute necessary for validation (it's assumed that validation against JSON schema happens only once, at the `Message.from_json` "level").

Does this make sense?